File: lib/mxflutter_framework/mxf_js_framework/packages/dio/dio.js
//  MXFlutterFramework
//  Copyright 2019 The MXFlutter Authors. All rights reserved.
//
//  Use of this source code is governed by a MIT-style license that can be
//  found in the LICENSE file.

const dart_sdk = require("dart_sdk");
const packages__dio__src__entry__dio_for_browser = require("packages/dio/src/entry/dio_for_browser.dart.lib.js");
const packages__dio__src__cancel_token = require("packages/dio/src/cancel_token.dart.lib.js");

function convertJSMap2DartMapOfString$dynamic(jsMap) {
    try {
        let es = Array();
        for (let key in jsMap) { // 遍历Map
            es.push(key);
            es.push(jsMap[key]);
        }
        return new (IdentityMapOfString$dynamic()).from(es);
    } catch (e$) {
        let e = dart.getThrown(e$);
        core.print("convertJSMap2DartMapOfString$dynamic error:" + e);
        return jsMap;

    }
}


//dart_sdk.dart.runtimeType(response.data).toString()

function convertDartJSONMap2jsMap(jsonMap) {
    try {
        if (dart_sdk.dart.runtimeType(jsonMap).toString() != "_JsonMap") {
            return jsonMap;
        }

        //TODO：优化
        let jsonStr = convert.json.encode(jsonMap);
        let jsMap = JSON.parse(jsonStr);
        return jsMap;
    } catch (e$) {
        let e = dart.getThrown(e$);
        core.print("convertDartJSONMap2jsMap error:" + e);
        return jsonMap;
    }
}


function checkIsLinkMap(pMap) {
    return pMap && typeof pMap === "LinkMap";
}

//mx_jsbuilder编译器使用的导出
exports.src__dio = packages__dio__src__entry__dio_for_browser.src__dio;
exports.src__response = packages__dio__src__cancel_token.src__response;
exports.src__options = packages__dio__src__cancel_token.src__options;
exports.src__cancel_token = packages__dio__src__cancel_token.src__cancel_token;
exports.src__dio_error = packages__dio__src__cancel_token.src__dio_error;

//JS直接开发使用的导出
const packages__mxdio = require("packages/dio/src/mxdio.js");
exports.Dio = packages__mxdio.MXDio;
exports.BaseOptions = packages__mxdio.MXBaseOptions
exports.Options = packages__mxdio.MXOptions;
exports.Headers = packages__mxdio.MXHeaders;
exports.Response = packages__mxdio.MXResponse;
exports.RequestOptions = packages__mxdio.MXRequestOptions
exports.ResponseType = packages__mxdio.MXResponseType;

File: mxflutter/lib/src/pkg/dio/src/dio.dart
//  MXFlutterFramework
//  Copyright 2019 The MXFlutter Authors. All rights reserved.
//
//  Use of this source code is governed by a MIT-style license that can be
//  found in the LICENSE file.

import 'package:mxflutter/src/mirror/mx_mirror.dart';
import 'package:dio/src/dio.dart';
import 'dart:async';
import 'dart:convert';
import 'dart:math' as math;
import 'dart:typed_data';
import 'package:dio/src/adapter.dart';
import 'package:dio/src/form_data.dart';
import 'package:dio/src/options.dart';
import 'package:dio/src/interceptor.dart';
import 'package:dio/src/headers.dart';
import 'package:dio/src/cancel_token.dart';
import 'package:dio/src/transformer.dart';
import 'package:dio/src/response.dart';
import 'package:dio/src/dio_error.dart';
import 'package:dio/src/entry/dio_for_native.dart';

///把自己能处理的类注册到分发器中
Map<String, MXFunctionInvoke> registerDioSeries() {
  var m = <String, MXFunctionInvoke>{};
  m[_dio.funName] = _dio;
  m[_dio_request.funName] = _dio_request;
  return m;
}

var _dio = MXFunctionInvoke(
  "Dio",
  ({
    BaseOptions options,
  }) =>
      Dio(
    options,
  ),
  [
    "options",
  ],
);

var _dio_request = MXFunctionInvoke(
  "Dio#request",
  ({
    Dio mirrorObj,
    String path,
    dynamic data,
    Map<String, dynamic> queryParameters,
    Options options,
    CancelToken cancelToken,
    dynamic onSendProgress,
    dynamic onReceiveProgress,
  }) async {
    Response response = await mirrorObj.request(
      path,
      data: data,
      queryParameters: queryParameters,
      options: options,
      cancelToken: cancelToken,
      // onSendProgress: createVoidTwoParamsClosure(_dio_request.buildOwner, onSendProgress),
      // onReceiveProgress: createVoidTwoParamsClosure(_dio_request.buildOwner, onReceiveProgress),
    );
    return _responseOBJtoJSON(response);
  },
  [
    "mirrorObj",
    "path",
    "data",
    "queryParameters",
    "options",
    "cancelToken",
    "onSendProgress",
    "onReceiveProgress",
  ],
);

String _responseOBJtoJSON(Response response) {
  var data = response.data;
  ResponseType respType = response.request.responseType;
  if (respType == ResponseType.bytes) {
    data = base64Encode(data);
  }
  Map jsonMap = {};
  jsonMap["data"] = response.data;
  jsonMap["headers"] = response.headers.map;
  jsonMap["isRedirect"] = response.isRedirect;
  jsonMap["statusCode"] = response.statusCode;
  jsonMap["statusMessage"] = response.statusMessage;
  //jsonMap["redirects"] = response.redirects;
  jsonMap["extra"] = response.extra;
  String sendStr = json.encode(jsonMap);
  return sendStr;
}
